doc/topics/async-system.md

```diff
@@ -0,0 +1,209 @@
+
+`AsyncSystem` gives us an elegant way to express this kind of sequential process involving a series of asynchronous steps.
+
+## Creating an AsyncSystem {#creating-an-asyncsystem}
+
+Most applications have a single `AsyncSystem` that is used throughout. It is constructed from an [ITaskProcessor](@ref CesiumAsync::ITaskProcessor) instance, which is a simple interface used to perform some work in a background thread. The simplest possible implementation looks like this:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp simplest-task-processor
+
+This implementation will work well, but it isn't very efficient because a brand new thread is created for each task. Most applications will implement this interface using a thread pool, task graph, or similar functionality that their application already contains.
+
+The `AsyncSystem` can then be created as follows:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp create-async-system
+
+`AsyncSystem` instances can be safely and efficiently stored and copied by value. This makes it easy to make them available wherever they're needed, including in lambda captures. You can think of an `AsyncSystem` instance as acting a bit like a smart pointer.
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp capture-by-value
+
+However, it is essential that the last `AsyncSystem` instance be destroyed only after everything is done using it. So a common pattern is to create and store it as a static local in an accessor function:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp async-system-singleton
+
+## Future<T> {#future}
+
+[Future<T>](@ref CesiumAsync::Future) represents an asynchronous operation that will eventually complete and produce a value of type `T`. If you're familiar with a JavaScript [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), the concept is very much the same. We commonly see `Future<T>` as the return value of a method that starts an asynchronous operation. For example, we can use [IAssetAccessor](@ref CesiumAsync::IAssetAccessor) to asynchronously download the content of a web URL:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp create-request-future
+
+This [get](@ref CesiumAsync::IAssetAccessor::get) function will start the HTTP request, but it will return quickly, without waiting for the request to complete. Initially, the `Future` is in the _Pending_ state, meaning that the asynchronous operation has not yet completed. Once it does complete, the `Future` will be in one of these two states:
+
+* _Resolved_: The operation has completed successfully and the result value is available.
+* _Rejected_: The operation failed, and an exception describing the error is available.
+
+If we want to block the current thread and wait for the request to complete, we can call [waitInMainThread](@ref CesiumAsync::Future::waitInMainThread):
```

Comment:
Do you mean `wait()` instead of `waitInMainThread()`? The concept of the main thread has only been touched on so far, and it seems you're talking here about the simplest way to wait on a `Future`'s results. And then there's the confusing subject of `waitInMainThread()` meaning that it can only be called in the main thread... :smiling_imp:

Reply:
Good idea. I changed to `wait`, but also added a brief note about its potential for deadlocking and the fact that you should use `waitInMainThread` if you need to wait in the main thread. By the way, it's not so much that `waitInMainThread` must be called in the main thread, more that it _defines_ the main thread, and that definition better not conflict with any other (such as another thread calling `dispatchMainThreadTasks`) or you'll have a problem.